File: app/src/main/java/com/ape/transfer/adapter/OldPhonePickupAdapter.java
package com.ape.transfer.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ape.backuprestore.PersonalItemData;
import com.ape.transfer.R;

import java.util.ArrayList;

/**
 * Created by android on 16-7-16.
 */
public class OldPhonePickupAdapter extends RecyclerView.Adapter<OldPhonePickupAdapter.ViewHolder> {

    private LayoutInflater mInflater;
    private ArrayList<PersonalItemData> mPersonalItemDatas;
    private OnItemClickListener mListener;

    public OldPhonePickupAdapter(Context context, OnItemClickListener onItemClickListener) {
        mInflater = LayoutInflater.from(context);
        setHasStableIds(true);
        mPersonalItemDatas = new ArrayList<>();
        mListener = onItemClickListener;
    }

    public void setDatas(ArrayList<PersonalItemData> datas) {
        mPersonalItemDatas = datas;
        notifyDataSetChanged();
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.item_data, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        PersonalItemData itemData = mPersonalItemDatas.get(position);
        holder.itemView.setTag(itemData);

        holder.ivIcon.setImageResource(itemData.getIconId());
        holder.tvName.setText(itemData.getTextId());
        holder.tvCount.setText(String.valueOf(itemData.getCount()));
        holder.ivSelected.setVisibility(itemData.isSelected() ? View.VISIBLE : View.GONE);
    }

    @Override
    public int getItemCount() {
        return mPersonalItemDatas.size();
    }

    @Override
    public long getItemId(int position) {
//        return super.getItemId(position);
        return mPersonalItemDatas.get(position).getType();
    }

    public interface OnItemClickListener {
        void onItemClick(View v);
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        ImageView ivIcon;
        ImageView ivSelected;
        TextView tvCount;
        TextView tvName;

        public ViewHolder(View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            ivIcon = (ImageView) itemView.findViewById(R.id.iv_icon);
            ivSelected = (ImageView) itemView.findViewById(R.id.iv_selected);
            tvCount = (TextView) itemView.findViewById(R.id.tv_count);
            tvName = (TextView) itemView.findViewById(R.id.tv_name);
        }

        @Override
        public void onClick(View v) {
            PersonalItemData itemData = (PersonalItemData) v.getTag();
            int index = mPersonalItemDatas.indexOf(itemData);
            boolean isSelected = itemData.isSelected();
            itemData.setSelected(!isSelected);
            notifyItemChanged(index);

            mListener.onItemClick(v);
        }
    }
}

File: app/src/main/java/com/ape/transfer/adapter/OldPhonePickupAdapter.kt
package com.ape.transfer.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView

import com.ape.backuprestore.PersonalItemData
import com.ape.transfer.R

import java.util.ArrayList

import butterknife.BindView
import butterknife.ButterKnife

/**
 * Created by android on 16-7-15.
 */
class OldPhonePickupAdapter(context: Context, private val mListener: OldPhonePickupAdapter.OnItemClickListener) : RecyclerView.Adapter<OldPhonePickupAdapter.ViewHolder>() {

    private var mBackupDataList: ArrayList<PersonalItemData>? = null
    private val mInflater: LayoutInflater

    init {
        mInflater = LayoutInflater.from(context)
        setHasStableIds(true)
        mBackupDataList = ArrayList<PersonalItemData>()
    }

    fun setDatas(backupDataList: ArrayList<PersonalItemData>) {
        mBackupDataList = backupDataList
        notifyDataSetChanged()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = mInflater.inflate(R.layout.item_data, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val itemData = mBackupDataList!![position]
        holder.tvName.setText(itemData.textId)
        holder.tvCount.text = itemData.count.toString()
        holder.ivIcon.setImageResource(itemData.iconId)
        holder.ivSelected.visibility = if (itemData.isSelected) View.VISIBLE else View.GONE
    }

    override fun getItemCount(): Int {
        return mBackupDataList!!.size
    }

    interface OnItemClickListener {
        fun onItemClick(v: View)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        internal var ivIcon: ImageView
        internal var ivSelected: ImageView
        internal var tvCount: TextView
        internal var tvName: TextView

        init {
            ButterKnife.bind(this, itemView)
            itemView.setOnClickListener(this)
            ivIcon = itemView.findViewById(R.id.iv_icon) as ImageView
            ivSelected = itemView.findViewById(R.id.iv_selected) as ImageView
            tvCount = itemView.findViewById(R.id.tv_count) as TextView
            tvName = itemView.findViewById(R.id.tv_name) as TextView

        }

        override fun onClick(v: View) {
            mListener.onItemClick(v)
        }
    }
}
